internal/http.go

```diff
@@ -0,0 +1,201 @@
+	}
+
+	w.WriteHeader(http.StatusOK)
+
+	_, err = w.Write([]byte("generated report"))
+	if err != nil {
+		logger.Error(err, "failed writing http response:")
+	}
+
+	return nil
+}
+
+func validateRequest(r *http.Request) *HTTPError {
+	if r.URL.Path != "/download" {
+		return &HTTPError{http.StatusNotFound, "404 not found", nil}
+	}
+
+	if r.Method != "GET" {
+		return &HTTPError{http.StatusMethodNotAllowed, "method is not supported", nil}
+	}
+
+	query := r.URL.Query()
+	start, present := query["start"]
+	if !present || len(start) == 0 {
+		return &HTTPError{http.StatusBadRequest, "start time not present", nil}
+	}
+
+	end, present := query["end"]
+	if !present || len(end) == 0 {
+		return &HTTPError{http.StatusBadRequest, "end time not present", nil}
+	}
+	return nil
+}
+
+func generateReport(ctx context.Context, conn *pgx.Conn, r *http.Request, logger logr.Logger) error {
+	query := r.URL.Query()
+	start := query["start"]
+	end := query["end"]
+	dirName := "/tmp/curator-report"
+	err := os.MkdirAll(dirName, 0755)
+	if err != nil {
+		logger.Error(err, "error making dir:")
+		return err
+	}
+	tarDirName := fmt.Sprintf("%s/%s-%s-koku-metrics.tar.gz",
+		dirName, strings.Join(start, ","), strings.Join(end, ","))
+	outputFile, err := os.OpenFile(tarDirName, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0644)
+	if err != nil {
+		logger.Error(err, "error opening file:")
+		return err
+	}
+	defer outputFile.Close()
+	writer, err := gzip.NewWriterLevel(outputFile, gzip.BestCompression)
+	if err != nil {
+		logger.Error(err, "error:")
+		return err
+	}
+	defer writer.Close()
+	tw := tar.NewWriter(writer)
+	defer tw.Close()
+	err = exporter(ctx, conn, tw, r, logger)
+	if err != nil {
+		return err
+	}
+	logger.Info("report generated successfully")
+	return nil
+}
+
+func exporter(ctx context.Context, conn *pgx.Conn, tw *tar.Writer, r *http.Request, logger logr.Logger) error {
+	query := r.URL.Query()
+	start := query["start"]
+	end := query["end"]
+	numberOfTables := 4
+	for i := 0; i < numberOfTables; i++ {
+		sqlQuery := fmt.Sprintf("COPY (SELECT * FROM logs_%d WHERE interval_start >= '%s'::timestamp with time zone AND interval_end < '%s' "+
+			"::timestamp with time zone) TO STDOUT WITH CSV DELIMITER ',' HEADER", i, strings.Join(start, ","), strings.Join(end, ","))
+		outfile := fmt.Sprintf("/tmp/curator-report/%s-%s-koku-metrics.%d.csv", strings.Join(start, ","), strings.Join(end, ","), i)
+		ef, err := os.OpenFile(outfile, os.O_RDWR|os.O_CREATE, 0755)
+		if err != nil {
+			logger.Error(err, "error opening file:")
+			return err
+		}
+		_, err = conn.PgConn().CopyTo(ctx, ef, sqlQuery)
+		if err != nil {
+			logger.Error(err, "error exporting report:")
+			return err
+		}
+
+		ef.Close()
+		err = addToArchive(tw, outfile)
+		if err != nil {
+			logger.Error(err, "error archiving file:")
+			return err
+		}
+		err = os.Remove(outfile)
+		if err != nil {
+			logger.Error(err, "error removing file:")
+			return err
```

Comment:
nit/not blocking: logging an error and then returning an error can be seen as anti-pattern. Just return a formatted error and let the call site determine whether it's worth logging.